Notebook 1, Module 1, Data and Data Management, CAS Applied Data Science, 2020-08-19, S. Haug, University of Bern. 

# 1. Data Management

Estimated study time is up to 2 hours. According to your background and how much you want to learn, you may need more or less. You are supposed to google, read manuals and chat with others during working through this notebook in order to benefit fully.

**Learning outcomes - after completion you**
- Know about data sources, types and formats (see lecture slides)
- Able to import and export data in Python
- Able to do simple things with dataframes in Python
- Know about data volumes, metadata and quality
- Able to plot histograms and scatter plots in Python (tomorrow)

**Documentation on Pandas DataFrame**
- Python: https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.html

## Outline

     0. About data management
     1. Getting used to Jupyter lab
     2. Import datasets into a Python dataframe
     3. Indexing on a dataframe
     4. Sorting 
     5. Filtering
     6. Exporting
     7. Missing and bad data
     8. Metadata
     9. Working on the filesystem


In [1]:
print("hello, world\n")

hello, world



### 0. About data management

Handling or managing data involves many steps and technologies at many levels. Data my be colleccted by sensors. It can be a camera, a temperature sensor, a telescope, a microscope (with a camera), a microphone, a particle detector etc. Normally the data is then digitised, maybe preprocessed and written to some media in a certain format, e.g. hard disk. This part of the data management is normally taken care of by engineers. 

Data may also be collected from all sorts of databases, so data already collected somehow. Time series of financial data, customers, passengers, facebook likes, twitter tweets etc. This is data which is normally already on a media with some interface for access, e.g. paper to be read by a camera, a file on youtube, a table on wikipedia etc. We will look at some ways to collect such data. Some programming and computer skills are needed to do so. It may be that this part of the data management is taken care of by specialised computer scientists, but it may also be expected from a data scientist to have these skills. 

Analysing data with statisitical and machine learning tools, requires that the data is colleceted, cleaned, prepared for the tools. This is very often a very large part of a data analytics project and a prerequisite. It may involve removing bad data, filter out redundant and noisy data, unify the formats and types, transform the data etc. 
Thus, a data scientist must be able to perform this part of the data management. This notebook shows the basic operations with Python pandas. With other tools the concepts and operations are very similar.

After the data analysis, after the extraction of information and the creation of knowledge, the data is often stored or archived for the future (if this seems cheaper than regenerating the data at a later point). In larger institutions this part of the data management may include educated librarians and others, not necessarily the data scientist.  


### 1. Getting used to Jupyter lab and computational notebooks

Jupyter lab is a "laboratory" where you can write rich text notebooks with executable code via your browser. There are several so-called kernels or computational back ends, i.e Python, R, Julia, bash etc can be supported. The text is written as Markdown. Latex is also supported (good for math). You can export the notebook in various formats, e.g. html. Everything can be done via the various tabs, however, the key shortcuts make you faster. 

*Useful key combinations*

- Shift+Enter or Control+Enter = Run cell
- Alt+Enter = Run cell and Insert new cell below

*Exercises (10 min)*
- Change and run this cell
- Add a new cell and execute some python statement in it
- Study the tabs in the menu of your jupyter (lab) notebook

In [2]:
# Write some Python here

print('Hallo Nico')

Hallo Nico


### 2. Import a dataset into a Python DataFrame

Pandas is a Python Module/Library for data analysis and manipulation. It has the data structure DataFrame which is quite powerful. Features

- DataFrame object for data manipulation with integrated indexing.
- Tools for reading and writing data between in-memory data structures and different file formats.
- Data alignment and integrated handling of missing data.
- Reshaping and pivoting of data sets.
- Label-based slicing, fancy indexing, and subsetting of large data sets.
- Data structure column insertion and deletion.
- Group by engine allowing split-apply-combine operations on data sets.
- Data set merging and joining.
- Hierarchical axis indexing to work with high-dimensional data in a lower-dimensional data structure.
- Time series-functionality: Date range generation and frequency conversion, moving window statistics, moving window linear regressions, date shifting and lagging.

The module is highly optimized for performance, with critical code paths written in Cython or C. Documentation here: https://pandas.pydata.org/pandas-docs/stable/api.html

For input/output (I/O) there are methods for reading EXCEL, SQL databases, HTML tables, clipboard, SAS, STATA etc.

In [3]:
import pandas

In [11]:
# import the python module pandas with the abbreviation pd
import pandas as pd
# Read the data in the iris.csv file into a dataframe
dataframe = pd.read_csv('iris.csv', dtype='str', names=['slength','swidth','plength','pwidth','species'])  #data type is a string (str), i.e. not converted into numbers
dataframe # print data

slength swidth plength pwidth         species
0       5.1    3.5     1.4    0.2     Iris-setosa
1       4.9    3.0     1.4    0.2     Iris-setosa
2       4.7    3.2     1.3    0.2     Iris-setosa
3       4.6    3.1     1.5    0.2     Iris-setosa
4       5.0    3.6     1.4    0.2     Iris-setosa
..      ...    ...     ...    ...             ...
145     6.7    3.0     5.2    2.3  Iris-virginica
146     6.3    2.5     5.0    1.9  Iris-virginica
147     6.5    3.0     5.2    2.0  Iris-virginica
148     6.2    3.4     5.4    2.3  Iris-virginica
149     5.9    3.0     5.1    1.8  Iris-virginica

[150 rows x 5 columns]

The notebook shows 50 rows per default. We can change this as we like:


In [14]:
pd.set_option('display.max_rows', 10) # or whatever number you like or None  

In [15]:
dataframe

slength swidth plength pwidth         species
0       5.1    3.5     1.4    0.2     Iris-setosa
1       4.9    3.0     1.4    0.2     Iris-setosa
2       4.7    3.2     1.3    0.2     Iris-setosa
3       4.6    3.1     1.5    0.2     Iris-setosa
4       5.0    3.6     1.4    0.2     Iris-setosa
..      ...    ...     ...    ...             ...
145     6.7    3.0     5.2    2.3  Iris-virginica
146     6.3    2.5     5.0    1.9  Iris-virginica
147     6.5    3.0     5.2    2.0  Iris-virginica
148     6.2    3.4     5.4    2.3  Iris-virginica
149     5.9    3.0     5.1    1.8  Iris-virginica

[150 rows x 5 columns]

In [16]:
dataframe.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 5 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   slength  150 non-null    object
 1   swidth   150 non-null    object
 2   plength  150 non-null    object
 3   pwidth   150 non-null    object
 4   species  150 non-null    object
dtypes: object(5)
memory usage: 6.0+ KB


Find out on wikipedia what this Iris data is about. 

### Hint

There are some ways to get help about modules and methods in Jupyter:

In [ ]:
dataframe. #(tab completion - type tab after the dot to see all methods)

Get help about this object, if there is any:

In [22]:
dataframe?

And, what you will probably use the most, the online package information and examples by googling.

### Exercise 1.1 (10 min)

Read the Swiss BAG Corona data into a dataframe and look at it. The URL is
'https://www.bag.admin.ch/dam/bag/de/dokumente/mt/k-und-i/aktuelle-ausbrueche-pandemien/2019-nCoV/covid-19-datengrundlage-lagebericht.xlsx.download.xlsx/200325_Datengrundlage_Grafiken_COVID-19-Bericht.xlsx'.

You will probably need to consult https://pandas.pydata.org/docs/reference/io.html in order to find out how to read a excel sheets into a pandas dataframe. 

In [28]:
# Write your code here
url = 'https://www.bag.admin.ch/dam/bag/de/dokumente/mt/k-und-i/aktuelle-ausbrueche-pandemien/2019-nCoV/covid-19-datengrundlage-lagebericht.xlsx.download.xlsx/200325_Datengrundlage_Grafiken_COVID-19-Bericht.xlsx'
df=pd.read_excel(url, header=6)
pd.set_option('display.max_rows', None)
df

Datum  Fallzahlen pro Tag  Fallzahlen pro Tag, kumuliert  \
0   2020-02-24                   1                              1   
1   2020-02-25                   1                              2   
2   2020-02-26                  10                             12   
3   2020-02-27                  10                             22   
4   2020-02-28                  10                             32   
5   2020-02-29                  13                             45   
6   2020-03-01                  11                             56   
7   2020-03-02                  31                             87   
8   2020-03-03                  33                            120   
9   2020-03-04                  61                            181   
10  2020-03-05                  62                            243   
11  2020-03-06                  73                            316   
12  2020-03-07                  49                            365   
13  2020-03-08                  69                            434   
14  2020-03-09                 192                            626   
15  2020-03-10                 212                            838   
16  2020-03-11                 334                           1172   
17  2020-03-12                 357                           1529   
18  2020-03-13                 433                           1962   
19  2020-03-14                 420                           2382   
20  2020-03-15                 328                           2710   
21  2020-03-16                1064                           3774   
22  2020-03-17                1088                           4862   
23  2020-03-18                1214                           6076   
24  2020-03-19                 836                           6912   
25  2020-03-20                1147                           8059   
26  2020-03-21                 693                           8752   
27  2020-03-22                 547                           9299   
28  2020-03-23                1464                          10763   
29  2020-03-24                1244                          12007   
30  2020-03-25                1072                          13079   
31  2020-03-26                1118                          14197   
32  2020-03-27                1308                          15505   
33  2020-03-28                 723                          16228   
34  2020-03-29                 433                          16661   
35  2020-03-30                1310                          17971   
36  2020-03-31                1137                          19108   
37  2020-04-01                1016                          20124   
38  2020-04-02                 878                          21002   
39  2020-04-03                 928                          21930   
40  2020-04-04                 486                          22416   
41  2020-04-05                 281                          22697   
42  2020-04-06                 925                          23622   
43  2020-04-07                 715                          24337   
44  2020-04-08                 613                          24950   
45  2020-04-09                 559                          25509   
46  2020-04-10                 309                          25818   
47  2020-04-11                 262                          26080   
48  2020-04-12                 219                          26299   
49  2020-04-13                 245                          26544   
50  2020-04-14                 425                          26969   
51  2020-04-15                 328                          27297   
52  2020-04-16                 321                          27618   
53  2020-04-17                 291                          27909   
54  2020-04-18                 144                          28053   
55  2020-04-19                  86                          28139   
56  2020-04-20                 277                          28416   
57  202

### 3. Indexing on a DataFrame

By methods:

In [24]:
dataframe.iat[149,1] # by integer numbers (row 149, column 1)

'3.0'

In [25]:
dataframe.at[149,'slength'] # by row number and column name

'5.9'

In [30]:
dataframe.iat[149,1]=10 # Assign a new value

Indexing a group/subset of the dataframe

In [31]:
sdf = dataframe.iloc[0:10,0:5]
print(sdf)

  slength swidth plength pwidth      species
0     5.1    3.5     1.4    0.2  Iris-setosa
1     4.9    3.0     1.4    0.2  Iris-setosa
2     4.7    3.2     1.3    0.2  Iris-setosa
3     4.6    3.1     1.5    0.2  Iris-setosa
4     5.0    3.6     1.4    0.2  Iris-setosa
5     5.4    3.9     1.7    0.4  Iris-setosa
6     4.6    3.4     1.4    0.3  Iris-setosa
7     5.0    3.4     1.5    0.2  Iris-setosa
8     4.4    2.9     1.4    0.2  Iris-setosa
9     4.9    3.1     1.5    0.1  Iris-setosa


**Important** 
When you assign a (subset) of dataframe to a new one like above, no copy is made. This means that if you change values of the new frame, also the orginal frame will be changed. If you want a copy, you need to use the copy method.

Indexing by column names...

In [34]:
sdf['pwidth']

0    0.2
1    0.2
2    0.2
3    0.2
4    0.2
5    0.4
6    0.3
7    0.2
8    0.2
9    0.1
Name: pwidth, dtype: object

Selecting (filtering) by column valuess_df = dataframe[dataframe['species']=='Iris-setosa']

In [37]:
s_df = dataframe[dataframe['pwidth']=="0.2"]

In [38]:
s_df

slength swidth plength pwidth      species
0      5.1    3.5     1.4    0.2  Iris-setosa
1      4.9    3.0     1.4    0.2  Iris-setosa
2      4.7    3.2     1.3    0.2  Iris-setosa
3      4.6    3.1     1.5    0.2  Iris-setosa
4      5.0    3.6     1.4    0.2  Iris-setosa
7      5.0    3.4     1.5    0.2  Iris-setosa
8      4.4    2.9     1.4    0.2  Iris-setosa
10     5.4    3.7     1.5    0.2  Iris-setosa
11     4.8    3.4     1.6    0.2  Iris-setosa
14     5.8    4.0     1.2    0.2  Iris-setosa
20     5.4    3.4     1.7    0.2  Iris-setosa
22     4.6    3.6     1.0    0.2  Iris-setosa
24     4.8    3.4     1.9    0.2  Iris-setosa
25     5.0    3.0     1.6    0.2  Iris-setosa
27     5.2    3.5     1.5    0.2  Iris-setosa
28     5.2    3.4     1.4    0.2  Iris-setosa
29     4.7    3.2     1.6    0.2  Iris-setosa
30     4.8    3.1     1.6    0.2  Iris-setosa
33     5.5    4.2     1.4    0.2  Iris-setosa
35     5.0    3.2     1.2    0.2  Iris-setosa
36     5.5    3.5     1.3    0.2  Iris-setosa
38     4.4    3.0     1.3    0.2  Iris-setosa
39     5.1    3.4     1.5    0.2  Iris-setosa
42     4.4    3.2     1.3    0.2  Iris-setosa
46     5.1    3.8     1.6    0.2  Iris-setosa
47     4.6    3.2     1.4    0.2  Iris-setosa
48     5.3    3.7     1.5    0.2  Iris-setosa
49     5.0    3.3     1.4    0.2  Iris-setosa

Vectorized dataframe manipulations...

In [35]:
sdf['half slength']=sdf['slength'].astype('float')/2.
sdf

slength swidth plength pwidth      species  half slength
0     5.1    3.5     1.4    0.2  Iris-setosa          2.55
1     4.9    3.0     1.4    0.2  Iris-setosa          2.45
2     4.7    3.2     1.3    0.2  Iris-setosa          2.35
3     4.6    3.1     1.5    0.2  Iris-setosa          2.30
4     5.0    3.6     1.4    0.2  Iris-setosa          2.50
5     5.4    3.9     1.7    0.4  Iris-setosa          2.70
6     4.6    3.4     1.4    0.3  Iris-setosa          2.30
7     5.0    3.4     1.5    0.2  Iris-setosa          2.50
8     4.4    2.9     1.4    0.2  Iris-setosa          2.20
9     4.9    3.1     1.5    0.1  Iris-setosa          2.45

### 4. Sorting on a DataFrame

...

In [36]:
sdf.sort_values(['slength','swidth'])

slength swidth plength pwidth      species  half slength
8     4.4    2.9     1.4    0.2  Iris-setosa          2.20
3     4.6    3.1     1.5    0.2  Iris-setosa          2.30
6     4.6    3.4     1.4    0.3  Iris-setosa          2.30
2     4.7    3.2     1.3    0.2  Iris-setosa          2.35
1     4.9    3.0     1.4    0.2  Iris-setosa          2.45
9     4.9    3.1     1.5    0.1  Iris-setosa          2.45
7     5.0    3.4     1.5    0.2  Iris-setosa          2.50
4     5.0    3.6     1.4    0.2  Iris-setosa          2.50
0     5.1    3.5     1.4    0.2  Iris-setosa          2.55
5     5.4    3.9     1.7    0.4  Iris-setosa          2.70

In [46]:
sdf["swidth"]=sdf["swidth"].astype("float") #change type of values in a dataframe

In [47]:
sdf.dtypes

slength     object
swidth     float64
plength     object
pwidth      object
species     object
dtype: object

In [37]:
sdf

slength swidth plength pwidth      species  half slength
0     5.1    3.5     1.4    0.2  Iris-setosa          2.55
1     4.9    3.0     1.4    0.2  Iris-setosa          2.45
2     4.7    3.2     1.3    0.2  Iris-setosa          2.35
3     4.6    3.1     1.5    0.2  Iris-setosa          2.30
4     5.0    3.6     1.4    0.2  Iris-setosa          2.50
5     5.4    3.9     1.7    0.4  Iris-setosa          2.70
6     4.6    3.4     1.4    0.3  Iris-setosa          2.30
7     5.0    3.4     1.5    0.2  Iris-setosa          2.50
8     4.4    2.9     1.4    0.2  Iris-setosa          2.20
9     4.9    3.1     1.5    0.1  Iris-setosa          2.45

You may measure the time needed for the execution, e.g. with the operating system command "time". System commands are executed with a %:

In [38]:
%time s_df[s_df['species']=='Iris-setosa'].sort_values('slength')

Wall time: 997 µs


slength swidth plength pwidth      species
13     4.3    3.0     1.1    0.1  Iris-setosa
8      4.4    2.9     1.4    0.2  Iris-setosa
38     4.4    3.0     1.3    0.2  Iris-setosa
42     4.4    3.2     1.3    0.2  Iris-setosa
41     4.5    2.3     1.3    0.3  Iris-setosa
22     4.6    3.6     1.0    0.2  Iris-setosa
3      4.6    3.1     1.5    0.2  Iris-setosa
47     4.6    3.2     1.4    0.2  Iris-setosa
6      4.6    3.4     1.4    0.3  Iris-setosa
2      4.7    3.2     1.3    0.2  Iris-setosa
29     4.7    3.2     1.6    0.2  Iris-setosa
24     4.8    3.4     1.9    0.2  Iris-setosa
30     4.8    3.1     1.6    0.2  Iris-setosa
12     4.8    3.0     1.4    0.1  Iris-setosa
11     4.8    3.4     1.6    0.2  Iris-setosa
45     4.8    3.0     1.4    0.3  Iris-setosa
9      4.9    3.1     1.5    0.1  Iris-setosa
37     4.9    3.1     1.5    0.1  Iris-setosa
34     4.9    3.1     1.5    0.1  Iris-setosa
1      4.9    3.0     1.4    0.2  Iris-setosa
40     5.0    3.5     1.3    0.3  Iris-setosa
35     5.0    3.2     1.2    0.2  Iris-setosa
43     5.0    3.5     1.6    0.6  Iris-setosa
26     5.0    3.4     1.6    0.4  Iris-setosa
25     5.0    3.0     1.6    0.2  Iris-setosa
49     5.0    3.3     1.4    0.2  Iris-setosa
4      5.0    3.6     1.4    0.2  Iris-setosa
7      5.0    3.4     1.5    0.2  Iris-setosa
23     5.1    3.3     1.7    0.5  Iris-setosa
46     5.1    3.8     1.6    0.2  Iris-setosa
44     5.1    3.8     1.9    0.4  Iris-setosa
39     5.1    3.4     1.5    0.2  Iris-setosa
19     5.1    3.8     1.5    0.3  Iris-setosa
17     5.1    3.5     1.4    0.3  Iris-setosa
21     5.1    3.7     1.5    0.4  Iris-setosa
0      5.1    3.5     1.4    0.2  Iris-setosa
32     5.2    4.1     1.5    0.1  Iris-setosa
28     5.2    3.4     1.4    0.2  Iris-setosa
27     5.2    3.5     1.5    0.2  Iris-setosa
48     5.3    3.7     1.5    0.2  Iris-setosa
16     5.4    3.9     1.3    0.4  Iris-setosa
31     5.4    3.4     1.5    0.4  Iris-setosa
10     5.4    3.7     1.5    0.2  Iris-setosa
20     5.4    3.4     1.7    0.2  Iris-setosa
5      5.4    3.9     1.7    0.4  Iris-setosa
36     5.5    3.5     1.3    0.2  Iris-setosa
33     5.5    4.2     1.4    0.2  Iris-setosa
18     5.7    3.8     1.7    0.3  Iris-setosa
15     5.7    4.4     1.5    0.4  Iris-setosa
14     5.8    4.0     1.2    0.2  Iris-setosa

In [48]:
%ls -l # list the file in your current directory with the linux command ls

 Volume in drive C is Windows-SSD
 Volume Serial Number is 4251-E1BA

 Directory of C:\Users\nicok\Desktop\CAS ADS\Module 1


 Directory of C:\Users\nicok\Desktop\CAS ADS\Module 1


 Directory of C:\Users\nicok\Desktop\CAS ADS\Module 1


 Directory of C:\Users\nicok\Desktop\CAS ADS\Module 1


 Directory of C:\Users\nicok\Desktop\CAS ADS\Module 1



File Not Found



 Directory of C:\Users\nicok\Desktop\CAS ADS\Module 1


 Directory of C:\Users\nicok\Desktop\CAS ADS\Module 1


 Directory of C:\Users\nicok\Desktop\CAS ADS\Module 1


 Directory of C:\Users\nicok\Desktop\CAS ADS\Module 1


 Directory of C:\Users\nicok\Desktop\CAS ADS\Module 1


 Directory of C:\Users\nicok\Desktop\CAS ADS\Module 1


 Directory of C:\Users\nicok\Desktop\CAS ADS\Module 1


 Directory of C:\Users\nicok\Desktop\CAS ADS\Module 1


 Directory of C:\Users\nicok\Desktop\CAS ADS\Module 1



### 5. Filtering on a DataFrame
...

In [49]:
sdf['slength'].astype('float')>5.0 # Returns a vector with booleans 

0     True
1    False
2    False
3    False
4    False
5     True
6    False
7    False
8    False
9    False
Name: slength, dtype: bool

In [50]:
# We can assign the filtering result to a new dataframe
sdf_large = sdf[sdf['slength'].astype('float')>5.0]

In [51]:
# And print it
sdf_large

slength  swidth plength pwidth      species
0     5.1     3.5     1.4    0.2  Iris-setosa
5     5.4     3.9     1.7    0.4  Iris-setosa

### 6. Missing or bad data

Datasets, in particular before they are "cleaned", may contain missing or wrongly formated values. There are DataFrame methods to deal with this:

- DataFrame.dropna([axis, how, thresh, …])	Remove missing values.
- DataFrame.fillna([value, method, axis, …])	Fill NA/NaN values using the specified method
- DataFrame.replace([to_replace, value, …])	Replace values given in to_replace with value.
- DataFrame.interpolate([method, axis, limit, …])	Interpolate values according to different methods.

Retrieving and cleaning data is often the most time consuming part in a data science project.

### 7. Exporting dataframes

DataFrame has several export methods. (html, hdf5, ascii, excel etc). Let's write our file to a text file in a csv format. 

In [52]:
sdf.to_csv('~/Desktop/myout.txt')

Now we can look at the myout.txt file with some regulare txt viewer.
The exporting method is very powerful and can account for many different options.

### 8. Metadata


Metadata is data about the data, e.g. when was it collected, under which conditions, calibration etc. 
Metadata is normally not part of the statistical data analysis, however, needed for understanding and reproducibilty. 

DataFrame is not really made for storing metadata (should be done separately), but one can add new attributes to a dataframe:


In [44]:
sdf.myinstrument_name = 'Binky'
sdf.myinstrument_name

'Binky'

### Exercise 2

Write down here some minimal metadata about the Iris dataset:

- column names
- dimensions (how many columns and rows?) Note: The dimensions of a numpy table or dataframe is often called the "shape". In this case the shape would be 150x5 (2-dimensional object with 150 rows and 5 columns)
- units (centimeter or meter, currency, etc.)
- author/source/time/software version
- Digital identifier if there are different versions of the dataset

In summary: Everything needed for reproduction!

### 9. Working on the filesystem with the os module

When managing large datasets, one often has to organise files in the file system. This includes finding them, moving and copying them, creating new folders/directories, renaming them etc. This can easily be done from Python with the os (operating system) module. 

Try to understand the following code and alter it to do something else. 

In [45]:
import os

new_dir = os.getcwd() +'/newnbdir' # Get the current directory into a string, add /newnbdir to the string
os.makedirs(new_dir) # Create the folder
files = os.listdir() # List the files in the current folder
nb_files = [] # Create an empty list
for file in files:
    if 'ipynb' in file and file[0]!='.':
        nb_files.append(file)
for file in nb_files:
    os.system('cp '+file+' '+new_dir+'/')

### Exercise 3

Read in the iris dataset from iris.csv into a dataframe. Set the values in column 1 in row 39, 49 and 100 to NaN (use the nan method from the numpy package). Then replacethe NaN values to the average value of the respective column. Depending on how you do it, this may be about 10 lines of Python code.

In [2]:
import pandas as pd
import numpy as np


In [10]:
data=pd.read_csv('iris.csv', names=['slength','swidth','plength','pwidth','species']) #creating dataframe
pd.set_option("display.max_rows", None)
data

slength  swidth  plength  pwidth          species
0        5.1     3.5      1.4     0.2      Iris-setosa
1        4.9     3.0      1.4     0.2      Iris-setosa
2        4.7     3.2      1.3     0.2      Iris-setosa
3        4.6     3.1      1.5     0.2      Iris-setosa
4        5.0     3.6      1.4     0.2      Iris-setosa
5        5.4     3.9      1.7     0.4      Iris-setosa
6        4.6     3.4      1.4     0.3      Iris-setosa
7        5.0     3.4      1.5     0.2      Iris-setosa
8        4.4     2.9      1.4     0.2      Iris-setosa
9        4.9     3.1      1.5     0.1      Iris-setosa
10       5.4     3.7      1.5     0.2      Iris-setosa
11       4.8     3.4      1.6     0.2      Iris-setosa
12       4.8     3.0      1.4     0.1      Iris-setosa
13       4.3     3.0      1.1     0.1      Iris-setosa
14       5.8     4.0      1.2     0.2      Iris-setosa
15       5.7     4.4      1.5     0.4      Iris-setosa
16       5.4     3.9      1.3     0.4      Iris-setosa
17       5.1     3.5      1.4     0.3      Iris-setosa
18       5.7     3.8      1.7     0.3      Iris-setosa
19       5.1     3.8      1.5     0.3      Iris-setosa
20       5.4     3.4      1.7     0.2      Iris-setosa
21       5.1     3.7      1.5     0.4      Iris-setosa
22       4.6     3.6      1.0     0.2      Iris-setosa
23       5.1     3.3      1.7     0.5      Iris-setosa
24       4.8     3.4      1.9     0.2      Iris-setosa
25       5.0     3.0      1.6     0.2      Iris-setosa
26       5.0     3.4      1.6     0.4      Iris-setosa
27       5.2     3.5      1.5     0.2      Iris-setosa
28       5.2     3.4      1.4     0.2      Iris-setosa
29       4.7     3.2      1.6     0.2      Iris-setosa
30       4.8     3.1      1.6     0.2      Iris-setosa
31       5.4     3.4      1.5     0.4      Iris-setosa
32       5.2     4.1      1.5     0.1      Iris-setosa
33       5.5     4.2      1.4     0.2      Iris-setosa
34       4.9     3.1      1.5     0.1      Iris-setosa
35       5.0     3.2      1.2     0.2      Iris-setosa
36       5.5     3.5      1.3     0.2      Iris-setosa
37       4.9     3.1      1.5     0.1      Iris-setosa
38       4.4     3.0      1.3     0.2      Iris-setosa
39       5.1     3.4      1.5     0.2      Iris-setosa
40       5.0     3.5      1.3     0.3      Iris-setosa
41       4.5     2.3      1.3     0.3      Iris-setosa
42       4.4     3.2      1.3     0.2      Iris-setosa
43       5.0     3.5      1.6     0.6      Iris-setosa
44       5.1     3.8      1.9     0.4      Iris-setosa
45       4.8     3.0      1.4     0.3      Iris-setosa
46       5.1     3.8      1.6     0.2      Iris-setosa
47       4.6     3.2      1.4     0.2      Iris-setosa
48       5.3     3.7      1.5     0.2      Iris-setosa
49       5.0     3.3      1.4     0.2      Iris-setosa
50       7.0     3.2      4.7     1.4  Iris-versicolor
51       6.4     3.2      4.5     1.5  Iris-versicolor
52       6.9     3.1      4.9     1.5  Iris-versicolor
53       5.5     2.3      4.0     1.3  Iris-versicolor
54       6.5     2.8      4.6     1.5  Iris-versicolor
55       5.7     2.8      4.5     1.3  Iris-versicolor
56       6.3     3.3      4.7     1.6  Iris-versicolor
57       4.9     2.4      3.3     1.0  Iris-versicolor
58       6.6     2.9      4.6     1.3  Iris-versicolor
59       5.2     2.7      3.9     1.4  Iris-versicolor
60       5.0     2.0      3.5     1.0  Iris-versicolor
61       5.9     3.0      4.2     1.5  Iris-versicolor
62       6.0     2.2      4.0     1.0  Iris-versicolor
63       6.1     2.9      4.7     1.4  Iris-versicolor
64       5.6     2.9      3.6     1.3  Iris-versicolor
65       6.7     3.1      4.4     1.4  Iris-versicolor
66       5.6     3.0      4.5     1.5  Iris-versicolor
67       5.8     2.7      4.1     1.0  Iris-versicolor
68       6.2     2.2      4.5     1.5  Iris-versicolor
69       5.6     2.5      3.9     1.1  Iris-versicolor
70       5.9     3.2      4.8     1.8  Iris-versicolor
71       6.1     2.8      4.0     1.3  Iris-v

In [22]:
data.loc[[39,49,100],"swidth"]=np.nan #setting data to NaN


In [30]:
#calculate mean and fill NaN values
mean_swidth=round(data["swidth"].mean(skipna=True),1)
print(mean_swidth)
data.fillna(value=mean_swidth)


3.0


slength  swidth  plength  pwidth          species
0        5.1     3.5      1.4     0.2      Iris-setosa
1        4.9     3.0      1.4     0.2      Iris-setosa
2        4.7     3.2      1.3     0.2      Iris-setosa
3        4.6     3.1      1.5     0.2      Iris-setosa
4        5.0     3.6      1.4     0.2      Iris-setosa
5        5.4     3.9      1.7     0.4      Iris-setosa
6        4.6     3.4      1.4     0.3      Iris-setosa
7        5.0     3.4      1.5     0.2      Iris-setosa
8        4.4     2.9      1.4     0.2      Iris-setosa
9        4.9     3.1      1.5     0.1      Iris-setosa
10       5.4     3.7      1.5     0.2      Iris-setosa
11       4.8     3.4      1.6     0.2      Iris-setosa
12       4.8     3.0      1.4     0.1      Iris-setosa
13       4.3     3.0      1.1     0.1      Iris-setosa
14       5.8     4.0      1.2     0.2      Iris-setosa
15       5.7     4.4      1.5     0.4      Iris-setosa
16       5.4     3.9      1.3     0.4      Iris-setosa
17       5.1     3.5      1.4     0.3      Iris-setosa
18       5.7     3.8      1.7     0.3      Iris-setosa
19       5.1     3.8      1.5     0.3      Iris-setosa
20       5.4     3.4      1.7     0.2      Iris-setosa
21       5.1     3.7      1.5     0.4      Iris-setosa
22       4.6     3.6      1.0     0.2      Iris-setosa
23       5.1     3.3      1.7     0.5      Iris-setosa
24       4.8     3.4      1.9     0.2      Iris-setosa
25       5.0     3.0      1.6     0.2      Iris-setosa
26       5.0     3.4      1.6     0.4      Iris-setosa
27       5.2     3.5      1.5     0.2      Iris-setosa
28       5.2     3.4      1.4     0.2      Iris-setosa
29       4.7     3.2      1.6     0.2      Iris-setosa
30       4.8     3.1      1.6     0.2      Iris-setosa
31       5.4     3.4      1.5     0.4      Iris-setosa
32       5.2     4.1      1.5     0.1      Iris-setosa
33       5.5     4.2      1.4     0.2      Iris-setosa
34       4.9     3.1      1.5     0.1      Iris-setosa
35       5.0     3.2      1.2     0.2      Iris-setosa
36       5.5     3.5      1.3     0.2      Iris-setosa
37       4.9     3.1      1.5     0.1      Iris-setosa
38       4.4     3.0      1.3     0.2      Iris-setosa
39       5.1     3.0      1.5     0.2      Iris-setosa
40       5.0     3.5      1.3     0.3      Iris-setosa
41       4.5     2.3      1.3     0.3      Iris-setosa
42       4.4     3.2      1.3     0.2      Iris-setosa
43       5.0     3.5      1.6     0.6      Iris-setosa
44       5.1     3.8      1.9     0.4      Iris-setosa
45       4.8     3.0      1.4     0.3      Iris-setosa
46       5.1     3.8      1.6     0.2      Iris-setosa
47       4.6     3.2      1.4     0.2      Iris-setosa
48       5.3     3.7      1.5     0.2      Iris-setosa
49       5.0     3.0      1.4     0.2      Iris-setosa
50       7.0     3.2      4.7     1.4  Iris-versicolor
51       6.4     3.2      4.5     1.5  Iris-versicolor
52       6.9     3.1      4.9     1.5  Iris-versicolor
53       5.5     2.3      4.0     1.3  Iris-versicolor
54       6.5     2.8      4.6     1.5  Iris-versicolor
55       5.7     2.8      4.5     1.3  Iris-versicolor
56       6.3     3.3      4.7     1.6  Iris-versicolor
57       4.9     2.4      3.3     1.0  Iris-versicolor
58       6.6     2.9      4.6     1.3  Iris-versicolor
59       5.2     2.7      3.9     1.4  Iris-versicolor
60       5.0     2.0      3.5     1.0  Iris-versicolor
61       5.9     3.0      4.2     1.5  Iris-versicolor
62       6.0     2.2      4.0     1.0  Iris-versicolor
63       6.1     2.9      4.7     1.4  Iris-versicolor
64       5.6     2.9      3.6     1.3  Iris-versicolor
65       6.7     3.1      4.4     1.4  Iris-versicolor
66       5.6     3.0      4.5     1.5  Iris-versicolor
67       5.8     2.7      4.1     1.0  Iris-versicolor
68       6.2     2.2      4.5     1.5  Iris-versicolor
69       5.6     2.5      3.9     1.1  Iris-versicolor
70       5.9     3.2      4.8     1.8  Iris-versicolor
71       6.1     2.8      4.0     1.3  Iris-v

### Exercise 4 

Print the list of Swiss cantons with the number of registered positive COVID19 tests since March. The list should be sorted so that the highest number occurs first. This should be doable with 3-10 lines of Python.


# End of today - remember to fill the mandatory form

Everyone has to fill this form by 10 pm : https://forms.gle/bzDpGLUieVnheZUi6.
Have a nice evening and see you tomorrow.